### PR TITLE
Lr/split validation

### LINKED ISSUE #3494: [DEPENDENCY_REFACTOR] - Split quorum proposal validation
### What is this task and why do we need to work on it?

The task is to split the quorum proposal validation. We want to split it into smaller parts, preferably the first part would preliminary checks allowing spoofed proposals to be early discarded.

### What work will need to be done to complete this task?

Analyse validation process and split it in logical parts.

### Are there any other details to include?

Use the preliminary checked proposal in Quorum Proposal Task in dependency tasks instead of non-validated proposal.

### What are the acceptance criteria to close this issue?

Spoofed proposals should no longer trigger quorum proposal dependency tasks.

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by lukaszrzasik**
Closes #3494

### This PR: 

Adds a new `HotShotEvent::QuorumProposalPreliminarilyValidated`. This event indicates that the received quorum proposal has been preliminarily validated. This means that the quorum proposal 
i. is not for an old view,
ii. has been correctly signed by the leader of the current view,
iii. includes a valid justify QC.
The `HotShotEvent::QuorumProposalPreliminarilyValidated` is received by the `QuorumProposalTaskState` instead of `HotShotEvent::QuorumProposalRecv`. It protects the `QuorumProposalTaskState` against spoofed quorum proposals.

Additionally:
i. Moves the signature check from `validate_proposal_safety_and_liveness` to `validate_proposal_view_and_certs`.
ii. Broadcasts `HighQcUpdated` instead of `UpdateHighQc` at the end of `handle_quorum_proposal_recv` because it is more accurate and saves us from a redundant shared consensus state update.

### This PR does not: 

This PR allows dependency tasks-based consensus to progress when flooded with spoofed quorum proposals but the progression is much slower than normal.

### Key places to review: 

crates/task-impls/src/helpers.rs
crates/task-impls/src/quorum_proposal_recv/handlers.rs
crates/task-impls/src/quorum_proposal/mod.rs

### How to test this PR:

All tests should pass.
The change has been tested with `demo-native` and one of the nodes sending spoofed quorum proposals. The consensus progresses correctly. Only the faulty leader's views timeout.

### CODE REVIEW

File: crates/task-impls/src/helpers.rs [hunk 1]
```diff
@@ -500,9 +486,18 @@ pub fn validate_proposal_view_and_certs<TYPES: NodeType>(
...
+    // There is a mistake here originating in the genesis leaf/qc commit. This should be replaced by:
+    //
+    //    proposal.validate_signature(&quorum_membership)?;
+    //
+    // in a future PR.
+    let proposed_leaf = Leaf::from_quorum_proposal(&proposal.data);
     ensure!(
-        view_leader_key == *sender,
-        "Leader key does not match key in proposal"
+        view_leader_key.validate(&proposal.signature, proposed_leaf.commit().as_ref()),
```
Comment (jparr721) [inline crates/task-impls/src/helpers.rs:499]: This `validate_signature` method exists now. Should we just use that?

File: crates/task-impls/src/events.rs [hunk 2]
```diff
@@ -191,6 +191,13 @@ pub enum HotShotEvent<TYPES: NodeType> {
 
     /// A new high_qc has been updated in `Consensus`.
     HighQcUpdated(QuorumCertificate<TYPES>),
+
+    /// A quorum proposal has been preliminarily validated.
+    /// The preliminary checks include:
+    /// 1. The proposal is not for an old view
+    /// 2. The proposal has been correctly signed by the leader of the current view
+    /// 3. The justify QC is valid
+    QuorumProposalPreliminarilyValidated(Proposal<TYPES, QuorumProposal<TYPES>>),
```
Comment (jparr721) [inline crates/task-impls/src/events.rs:200]: Do you think it'd be helpful to elucidate how this event is different from the `QuorumProposalValidated` event? Unfortunately, it looks like the `QuorumProposalValidated` event's doc comment doesn't give much information right now, so an unfamiliar reader might not be able to trivially differentiate the two events without tracing the validation paths.

**Review by jparr721 — COMMENTED** [review]
Great work, just a couple things I'd like your opinion on.

**Review by bfish713 — DISMISSED** [review]

File: crates/task-impls/src/events.rs [hunk 3]
```diff
@@ -84,6 +84,11 @@ pub enum HotShotEvent<TYPES: NodeType> {
     /// All dependencies for the quorum vote are validated.
     QuorumVoteDependenciesValidated(TYPES::Time),
     /// A quorum proposal with the given parent leaf is validated.
+    /// The full validation checks include:
```
Comment (jparr721) [inline crates/task-impls/src/events.rs:87]: Love this!

File: crates/task-impls/src/helpers.rs [hunk 4]
```diff
@@ -499,11 +484,14 @@ pub fn validate_proposal_view_and_certs<TYPES: NodeType>(
...
     );
 
-    let view_leader_key = quorum_membership.leader(view);
-    ensure!(
-        view_leader_key == *sender,
-        "Leader key does not match key in proposal"
-    );
+    // Validate the proposal's signature. This should also catch if the leaf_commitment does not equal our calculated parent commitment
+    //
+    // There is a mistake here originating in the genesis leaf/qc commit. This should be replaced by:
```
Comment (jparr721) [inline crates/task-impls/src/helpers.rs:489]: Looks like we might be able to remove this comment entirely? Looks like we can keep the above, but the "there is a mistake here" part seems to be rectified.

**Review by jparr721 — COMMENTED** [review]
Looks good, one last little thing

**Review by jparr721 — APPROVED** [review]

**Pull request merged**